crypto/src/lib.rs

```diff
@@ -388,6 +406,109 @@ impl PublicKeyInner {
     }
 }
 
+/// `PublicKey` will be lazily deserialized inside WASM.
+/// This is needed for performance reasons, since `PublicKeyInner::from_bytes` is quite slow.
+/// However inside WASM in most cases `PublicKey` is used only for comparisons (==).
+/// See https://github.com/hyperledger/iroha/issues/5038 for details.
+#[cfg(target_family = "wasm")]
+mod lazy {
+    use alloc::{boxed::Box, vec::Vec};
+    use core::{borrow::Borrow, cell::OnceCell, fmt};
+
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    use crate::{Algorithm, PublicKeyInner};
+
+    #[derive(Clone, Eq)]
+    pub struct PublicKeyLazy {
+        algorithm: Algorithm,
+        payload: Vec<u8>,
+        inner: OnceCell<Box<PublicKeyInner>>,
+    }
+
+    impl PublicKeyLazy {
+        pub fn new(inner: PublicKeyInner) -> Self {
+            Self {
+                algorithm: inner.algorithm(),
+                payload: inner.payload(),
+                inner: OnceCell::from(Box::new(inner)),
+            }
+        }
+
+        pub fn new_lazy(algorithm: Algorithm, payload: Vec<u8>) -> Self {
+            Self {
+                algorithm,
+                payload,
+                inner: OnceCell::new(),
+            }
+        }
+
+        fn get_inner(&self) -> &PublicKeyInner {
+            self.inner.get_or_init(|| {
+                let inner = PublicKeyInner::from_bytes(self.algorithm, &self.payload)
+                    .expect("Public key deserialization at WASM side must not fail because data received from host side");
+                Box::new(inner)
+            })
+        }
+
+        pub fn algorithm(&self) -> Algorithm {
+            self.algorithm
+        }
+
+        pub fn to_raw(&self) -> (Algorithm, Vec<u8>) {
+            (self.algorithm, self.payload.clone())
+        }
+    }
+
+    impl Borrow<PublicKeyInner> for PublicKeyLazy {
+        fn borrow(&self) -> &PublicKeyInner {
+            self.get_inner()
+        }
+    }
+
+    impl fmt::Display for PublicKeyLazy {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            fmt::Display::fmt(self.get_inner(), f)
+        }
+    }
+
+    impl fmt::Debug for PublicKeyLazy {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            fmt::Debug::fmt(self.get_inner(), f)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for PublicKeyLazy {
```

Comment:
AFAIU, with this impl, the `PublicKey` will have a lazy representation when it's coming from a scale representation, but it will still be a full representation when it comes from json or string. I would prefer for it to always be lazy in wasm, regardless the source, as this distinction may become an unwelcome surprise down the line..